Rework parallel build and split

gezalore opened the pull request:
Still WIP due to the issues below, but this is the baseline for #2360 and also implements some of #2247. It works fine for the test suite  but a couple of issues remain for EH1/EH2:

- My recent changes to how trace routines are built means that without  `--output-split-ctrace` there is one huge trace function which remains the long pole in the build time. A similar issue remains for EH2, where there is a single sequent block that is 77k lines long and remains the long pole (and it is a very long pole). 

Question: How would you feel about defaulting `--output-split-cfuncs` and `--output-split-ctrace` to the value of `--output-split`, or some multiple thereof? Without something like that,  this PR has less then hoped for benefit for big builds.

- I think default 10000 is probably too small. It yields about 3000-4000 line files. I will investigate a bit later, but SweRV EH2 will have 40+ files (assuming we do the split-cfuncs above) which I feel might be too much, though on a modern servwer you can easily get 40+ cores to compiles it alli one go. Let me know if you have an oppinion on this.

- EH1 baseline speed is 9% slower with this patch (clang 10). This is completely unexpected, but I have seen similar amount of slowdown from trivial code changes I made perviously, so I think this is hitting some behaviour threshold possibly around inlining in clang. I know for a fact tha `_eval` in EH1 has a lot more branches with smaller sub-functions than EH2, and EH2 speed is the same.

Question: Would you be happy to commit this with the performance regression, while raising a new issue to investigate why output split slows down EH1? Otherwise will have to dig into that right now which might take a while to resolve.

gezalore commented:
One more data point on model peformance (EH1 coremark run-time in seconds):

| | Clang 10.0 | GCC 9.3 |
|-|-|-|
| before this PR | 8.88 | 11.90|
| with this PR | 9.70 | 11.5 1|

It is not a clear cut case.

wsnyder commented on include/verilated.mk.in:
```diff
@@ -163,35 +157,36 @@ ifneq ($(VK_LIBS_THREADED),0)
-# Overall Objects Linking
+### Overall Objects Linking
 
-VK_CLASSES_FAST_CPP = $(addsuffix .cpp, $(VM_CLASSES_FAST))
-VK_CLASSES_SLOW_CPP = $(addsuffix .cpp, $(VM_CLASSES_SLOW))
-
-VK_SUPPORT_FAST_CPP = $(addsuffix .cpp, $(VM_SUPPORT_FAST))
-VK_SUPPORT_SLOW_CPP = $(addsuffix .cpp, $(VM_SUPPORT_SLOW))
```
What you did is cleaner, but I'd prefer to keep any of the old defines, then use both VK_CLASSES/SUPPORT_FAST/SLOW below. People may have custom make files that would then break otherwise (I have a few such makefiles - though might be the only one).

I'm not worried about the ALLfast/slow being renamed ALL though.

gezalore replied:
I appreciate that situation, I thougth that giventhese are only used under `ifneq ($(VM_PARALLEL_BUILDS),1)` they are not that significant. I'm happy to keep the old one though would love to know what you use them for externally (just curious, so don't spend time answering this if it's complicated).

wsnyder replied:
Let's cleanup then. As you had it.  Looking at the makefile I had it has other issues and this is a good excuse to fix it the right way.

wsnyder commented on include/verilated.mk.in:
```diff
@@ -202,19 +197,15 @@ $(VM_PREFIX)__ALL.a: $(VK_OBJS)
 	$(OBJCACHE) $(CXX) $(CXXFLAGS) $(CPPFLAGS) $(OPT_FAST) -c -o $@ $<
 
-%__Slow.o: %__Slow.cpp
+$(VK_SLOW_OBJS): %.o: %.cpp
```
If someone has any user files named __Slow this will change behavior.  I also sort of like having a name that I can be sure compiles a certain way.  Quite open to counter-argument.

gezalore replied:
I did this because __Syms.cpp is a slow file but the old rule doesn't aply to that.

Options are:
1 Don't care, keep building __Syms.cpp as fast.
2 Emit as __Syms__Slow.cpp 
3 Keep new rule but add old rule as well for backwards compatibility
4 Keep old rule, add new rule just for __Syms.cpp (will break if we add another slow file that is not *__Slow.cpp)

I appreciate backwards compatibility but at the same time, I don't think this is a documented feature, so in the best case people are just piggy backing on it. I personally don't like filenames having any meaning but a human readable name as I never remember what substring is magic in who's build system+ also might have to do with me not being able to type, though that is not much of an argument :)

I also did it this way because I was planning to add a 3rd rule later with OPT_GLOBAL (based on what we discussed in #2247) to use for the run-time library bits, as for a speedy debug cyle one might want OPT_FAST=-O0, but the run-time lib is tiny in comparison and contains all tracing routines so OPT_FAST=-O0 and OPT_GLOBAL=-O2 would be way faster with tracing at insignificant build time cost (especially that the GLOBAL files ccache well). That new rule would look the same but with `s/SLOW/GLOBAL/g`.

wsnyder replied:
Ok with how you did it now that I understand, thanks.

wsnyder commented:
>Question: How would you feel about defaulting --output-split-cfuncs and --output-split-ctrace to the value of --output-split, or some multiple thereof? Without something like that, this PR has less then hoped for benefit for big builds.

I think in V3Options::notify you can set each to the minimum of their value or --output-split presuming output-split != 0.  I'd like to be able to make them smaller separately for debug.

>I think default 10000 is probably too small. It yields about 3000-4000 line files.

15000 is actually what worked good on EL1 on an 8-core machine. Maybe try that?

>    EH1 baseline speed is 9% slower with this patch (clang 10).

Ick.  Assuming 15000 is similar I don't think we can commit this part (output-split-on) without understanding that.  Again, we should separate out PARALLEL/ALL as that won't affect this.

BTW note on EH1 with earlier GCC I got a BOOST of a few percent as some gcc optimizations were off on huge funcs.

gezalore commented:
> I think in V3Options::notify you can set each to the minimum of their value or --output-split presuming output-split != 0. I'd like to be able to make them smaller separately for debug.

What I have done is set them to the value of output-split, unless the option is explicitly specified, in which case use whatever value the usere provided, isn't that preferable? It's earier to think about,but yes setting it to a way larger than output-split is unwise.

> BTW note on EH1 with earlier GCC I got a BOOST of a few percent as some gcc optimizations were off on huge funcs

Note I do get a small boost (3.4%) with GCC, the regression is with clang only.

Measurements for an 8 core host follow. These are with setting --output-split-cfuncs and --output-split-ctrace to the same as --output-split):

```
Clang 10

EH1:
  ccache -C -z; time make -j8 -f $RV_ROOT/tools/Makefile VERILATOR=$VERILATOR \
  CONF_PARAMS=-iccm_enable=1 TEST=cmark_iccm \
  VERILATOR_MAKE_FLAGS="OPT_FAST='-O2 -g' OPT_SLOW='-g'" verilator-build program.hex

EH2:
  ccache -C -z; time make -j8 -f $RV_ROOT/tools/Makefile VERILATOR=$VERILATOR \
  CONF_PARAMS=-iccm_enable=1 target=default_mt TEST=cmark_iccm_mt \
  VERILATOR_MAKE_FLAGS="OPT_FAST='-O2 -g' OPT_SLOW='-g'" verilator-build program.hex

EL2:
  ccache -C -z; time make -j8 -f $RV_ROOT/tools/Makefile VERILATOR=$VERILATOR \
  CONF_PARAMS=-iccm_enable=1 TEST=cmark_iccm \
  VERILATOR_MAKE_FLAGS="OPT_FAST='-O2 -g' OPT_SLOW='-g'" verilator-build program.hex

          |          EH1          |          EH2          |          EL2          |
          |  cpu  | wall  |  run  |  cpu  | wall  |  run  |  cpu  | wall  |  run  |
==========|=======|=======|=======|=======|=======|=======|=======|=======|=======|
split 0   | 129.5 | 124.4 | 8.814 | 509.4 | 506.0 | 28.33 | 124.5 | 120.7 | 9.866 |
split 10k | 175.6 |  28.6 | 9.511 | 238.1 |  54.0 | 27.58 |  87.9 |  16.8 | 9.810 |
split 15k | 158.7 |  28.8 | 9.383 | 206.8 |  49.5 | 27.46 |  77.8 |  15.4 | 9.600 |
split 20k | 140.0 |  30.8 | 9.310 | 188.0 |  46.5 | 27.68 | 136.1 |  71.7 | 9.809 |
split 30k | 120.0 |  30.4 | 9.282 | 287.0 | 140.9 | 28.01 |  68.3 |  14.1 | 9.862 |
```
The sweet spot does semm to be around 15-20k. I would lean towards 20k as I think too small is more harmful to performance than too lartge. Note the distinct anomaly for EL2 with 20k which I will investigate before concluding this.

I think most of the performance regression in EH1 on clang is due to inlining. Wit inlining turned off I get this:

```
EH1 without inlining:
  ccache -C -z ; time make -j8 -f $RV_ROOT/tools/Makefile VERILATOR=$VERILATOR \
  CONF_PARAMS=-iccm_enable=1 TEST=cmark_iccm \
  VERILATOR_MAKE_FLAGS="OPT_FAST='-O2 -fno-inline -g' OPT_SLOW='-g -fno-inline'" \
  verilator-build program.hex

          |          EH1          |
          |  cpu  | wall  |  run  |
==========|=======|=======|=======|
split 0   |  97.4 |  93.2 | 9.642 |
split 10k | 134.9 |  26.2 | 9.830 |
split 15k | 114.6 |  24.4 | 9.727 |
split 20k | 106.3 |  22.9 | 9.718 |
split 30k |  95.3 |  22.4 | 9.715 |
```

In particular, EH1 has lots of Active blocks with 1 to 4 statements in them tops, we should just inline these into `_eval`.

EH1 `_eval` actually has stuff like this in it which makes me wonder if ther is a bug in working out the sensitivities somewhere as I doubt the folks at WD actually designed this (note: this is one of the medium sized ones):

```
    if (((((((((((((((((((((((((((((((((((~ (IData)(vlTOPp->__VinpClk__TOP__tb_top__DOT__rvtop__DOT__core_rst_l)) 
                                         & (IData)(vlTOPp->__Vclklast__TOP____VinpClk__TOP__tb_top__DOT__rvtop__DOT__core_rst_l)) 
                                        | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__active_clk) 
                                           & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__active_clk)))) 
                                       | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__dbg__DOT__dbg_data0_reg__DOT__l1clk) 
                                          & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__dbg__DOT__dbg_data0_reg__DOT__l1clk)))) 
                                      | ((~ (IData)(vlTOPp->__VinpClk__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__dbg__DOT__dbg_dm_rst_l)) 
                                         & (IData)(vlTOPp->__Vclklast__TOP____VinpClk__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__dbg__DOT__dbg_dm_rst_l))) 
                                     | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__decode__DOT__trap_e1ff__DOT__l1clk) 
                                        & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__decode__DOT__trap_e1ff__DOT__l1clk)))) 
                                    | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__decode__DOT__trap_e2ff__DOT__l1clk) 
                                       & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__decode__DOT__trap_e2ff__DOT__l1clk)))) 
                                   | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__decode__DOT__trap_e3ff__DOT__l1clk) 
                                      & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__decode__DOT__trap_e3ff__DOT__l1clk)))) 
                                  | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__decode__DOT__trap_e4ff__DOT__l1clk) 
                                     & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__decode__DOT__trap_e4ff__DOT__l1clk)))) 
                                 | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__decode__DOT__wbff__DOT__l1clk) 
                                    & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__decode__DOT__wbff__DOT__l1clk)))) 
                                | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__instbuff__DOT__cinst0ff__DOT__l1clk) 
                                   & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__instbuff__DOT__cinst0ff__DOT__l1clk)))) 
                               | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__instbuff__DOT__cinst1ff__DOT__l1clk) 
                                  & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__instbuff__DOT__cinst1ff__DOT__l1clk)))) 
                              | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__tlu__DOT__mcgc_ff__DOT__l1clk) 
                                 & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__tlu__DOT__mcgc_ff__DOT__l1clk)))) 
                             | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__tlu__DOT__mfdc_ff__DOT__l1clk) 
                                & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__dec__DOT__tlu__DOT__mfdc_ff__DOT__l1clk)))) 
                            | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__dma_ctrl__DOT__dma_buffer_c1_clk) 
                               & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__dma_ctrl__DOT__dma_buffer_c1_clk)))) 
                           | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__dma_ctrl__DOT__dma_free_clk) 
                              & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__dma_ctrl__DOT__dma_free_clk)))) 
                          | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__free_clk) 
                             & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__free_clk)))) 
                         | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__dccm_ctl__DOT__picm_data_ff__DOT__l1clk) 
                            & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__dccm_ctl__DOT__picm_data_ff__DOT__l1clk)))) 
                        | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_bus_buf_c1_clk) 
                           & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_bus_buf_c1_clk)))) 
                       | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_c1_dc3_clk) 
                          & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_c1_dc3_clk)))) 
                      | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_c1_dc4_clk) 
                         & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_c1_dc4_clk)))) 
                     | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_c1_dc5_clk) 
                        & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_c1_dc5_clk)))) 
                    | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_c2_dc3_clk) 
                       & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_c2_dc3_clk)))) 
                   | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_c2_dc4_clk) 
                      & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_c2_dc4_clk)))) 
                  | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_c2_dc5_clk) 
                     & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_c2_dc5_clk)))) 
                 | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_dccm_c1_dc3_clk) 
                    & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_dccm_c1_dc3_clk)))) 
                | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_free_c2_clk) 
                   & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_free_c2_clk)))) 
               | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_freeze_c1_dc2_clk) 
                  & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_freeze_c1_dc2_clk)))) 
              | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_freeze_c1_dc3_clk) 
                 & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_freeze_c1_dc3_clk)))) 
             | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_freeze_c2_dc1_clk) 
                & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_freeze_c2_dc1_clk)))) 
            | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_freeze_c2_dc2_clk) 
               & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_freeze_c2_dc2_clk)))) 
           | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_freeze_c2_dc3_clk) 
              & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_freeze_c2_dc3_clk)))) 
          | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_lsc_ctl__DOT__rs1ff__DOT__l1clk) 
             & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_lsc_ctl__DOT__rs1ff__DOT__l1clk)))) 
         | ((IData)(vlTOPp->tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_stbuf_c1_clk) 
            & (~ (IData)(vlTOPp->__Vclklast__TOP__tb_top__DOT__rvtop__DOT__swerv__DOT__lsu__DOT__lsu_stbuf_c1_clk))))) {
        vlTOPp->_multiclk__TOP__880(vlSymsp);
        vlTOPp->__Vm_traceActivity[0x12dU] = 1U;
    }
```

I was kind of on the way to investigate this when I got sidetracked with #2360 to start with. I think the rest of the performance loss with spliting is because Clang could probably do some interporcedural analysis and eliminate some of these conditions while with black box functions it can't. EH2/EL2 doesn't have these.

gezalore commented:
> Please split out the VM_PARALLEL_BUILDS and ALL fix as that should be a slam-dunk versus the performance issue. (Likewise if someone tries to bisect a performance gain/loss.)

Sorry I'm not following. This PR has 2 patches which sholdn't be squashed. The first one fuses the ALL files to one, the second one sets --output-split by default an VM_PARALLEL_BUILDS=1 only if code size > split size. Is this what you mean?

wsnyder commented:
>What I have done is set them to the value of output-split, unless the option is explicitly specified, in which case use whatever value the usere provided, isn't that preferable? 

Ok.

>Note I do get a small boost (3.4%) with GCC, the regression is with clang only.

That makes me willing to take it them.

>The sweet spot does semm to be around 15-20k. I would lean towards 20k as I think too small is more harmful to performance than too lartge. Note the distinct anomaly for EL2 with 20k which I will investigate before concluding this.

Ok.

>In particular, EH1 has lots of Active blocks with 1 to 4 statements in them tops, we should just inline these into _eval.

Yup, another pull to do!

>EH1 _eval actually has stuff like this in it which makes me wonder if ther is a bug in working out the sensitivities somewhere as I doubt the folks at WD actually designed this (note: this is one of the medium sized ones):

define RV_FPGA_OPTIMIZE.  They were supposed to fix that before the release but seems got lost. Let me know if that fixes it, I will file a bug on them.

>> Please split out the VM_PARALLEL_BUILDS and ALL fix as that should be a slam-dunk versus the performance issue. (Likewise if someone tries to bisect a performance gain/loss.)
>Sorry I'm not following. This PR has 2 patches which sholdn't be squashed. The first one fuses the ALL files to one, the second one sets --output-split by default an VM_PARALLEL_BUILDS=1 only if code size > split size. Is this what you mean?

Would like three then.  First (ALL merge) seems fine once comments are addressed.  Second would set VM_PARALLEL_BUILDS appropriately - everything but change output split default. Third changes output split default.  Thus a problem if bisected would see only the change of the default (& any test changes fallout) not the other changes.

wsnyder commented on src/V3Global.h:
```diff
@@ -129,6 +131,12 @@ class V3Global {
     void dpi(bool flag) { m_dpi = flag; }
+    uint64_t codeSize() const { return m_codeSize; }
+    void codeSizeInc(uint64_t increment) { m_codeSize += increment; }
+    bool useParallelBuild() const {
+        const uint64_t limit = static_cast<uint64_t>(opt.outputSplit());
+        return limit > 0 && codeSize() > limit;
+    }
```
I understand the files might be e.g. 4x bigger, though suspect non-linearness is unlikely to matter much between the separate outputs as they don't have much in common. Anyhow I'd really prefer the rule I proposed as it's easily visible from the output if correct or not  basically if we see any __1 file (a split) we know it will parallel.

What I meant is we need a negative test - that is if you have a bug that sets VM_PARALLEL = 1 always, is there some obvious test_regress test that will fail?  (You could do that experiment to find out).  One option would be to copy t_flag_csplit.pl, take away the --output-split settings, and check what test objects get made - should be just one.

gezalore replied:
OK, I'm with you. Please see tests in middle commit. I just pushed.

gezalore commented:
Indeed RV_FPGA_OPTIMIZE helps a lot (though the ICGs under verilator are already just negedge sensitive rather than latches, so I wonder why it helps so much). Ther are still blocks with 20+ clocks, and 1 [rather complicaed] statement inside.

```
EH1 with +define+RV_FPGA_OPTIMIZE=1

          |          EH1          |
          |  cpu  | wall  |  run  |
==========|=======|=======|=======|
split 0   | 119.3 | 114.2 | 6.472 |
split 10k | 159.6 |  28.5 | 6.733 |
split 15k | 143.1 |  25.8 | 6.711 |
split 20k | 139.0 |  29.0 | 6.636 |
split 30k | 115.3 |  33.2 | 6.611 |
```

gezalore commented:
I split out to have the 3 commits.

I am still investing the EL2 built time anomaly, but I think once that is resovled this will be ready, as I think we understand the reason behind the EH1 clang speed regression.

gezalore commented:
BTW, set the default ot 20k

wsnyder commented:
I'm happy with it. Push non-squash when Travis & you are happy with it!

gezalore commented:
The EL2 build time anomaly seems to be hitting some algorithmic issue in LLVM's jump threading pass and as such is not an issue with this PR. I will file a separate ticket regarding this.

gezalore commented:
I will push this when Travis is happy.
